### DISCUSSION TITLE
Proposal: Add JSONFields to various django-CMS models

### DISCUSSION THREAD

**Discussion by jrief** [django-cms/django-cms#7074] [General]
## Preamble

In django-CMS many models can be extended by a third party Django-CMS app.
These models are `cms.models.pagemodel.Page`, `cms.models.titlemodels.Title`,
`cms.models.pluginmodels.CMSPlugin`, `filer.models.filemodels.File`.

With extendible I mean, that an implementor of django-CMS may inherit from
those models and create a One-to-One relationship between his model
implementation and the given base implementation as provided by the
django-CMS framework.

This approach is very powerful since it let's third-party implementors,
add arbitrary fields to existing models.

### Example:

An implementor of Django-CMS wants to add an icon to each of its pages.
This can be useful when building the menu tree, pointing onto these pages.
Then each menu entry may show an icon just before the menu title.

Up to now, the implementor must create a Django model inheriting from
`cms.extensions.models.PageExtension` and add a field to access that icon.
This per se is not a problem, but for each page added to the CMS, we now
must perform an additional database `JOIN` operation. The implementor, also
must take care of creating a function to properly deep copy the data
contained in the implementation of his model inheriting from `PageExtension`.

**Having to manage all these function is cumbersome and error-prone.**

## CMSPlugin

The most ovious model django-CMS offers for being extended by a third party
app, is the django-CMS plugin. Here our or other third party apps extends the model
`cms.models.pluginmodels.CMSPlugin` implemented by django-CMS. Usually the fields
in those models correspond to some HTML element tag names and attributes. They then
are rendered using a hard-coded HTML template tied to the plugin implementation.

By using a Django model, the django-admin backend creates an editor for us, so
that the CMS administrators can edit those fields. This is one of the reasons
for its success: It's just too simple to create a plugin with arbitrary fields
and immediatly having an editor to edit those fields. However, this simplicity
has some major drawbacks.

We often have many different plugin types but only a few instances of thereof.
This means that we have many database tables with only a few rows. This makes the
handling of the database difficult, since we need many migrations. Often, CMS
plugins contain many fields, from which the major part are optional. This means
that many columns in the tables reflecting the Django models remain empty.

**Therefore we really should rethink, if a relational database system is the
appropriate tool to keep unstructured data.**

From my point of view, many of those problems can be solved by keeping unstructured
data inside a JSONField, rather than in discrete database columns. We can drastically
reduce the number of database migrations. It also makes polymorphism of CMS plugins
much simpler.

Fortunately Django version 3.1 introduced a database independent JSONField, which
could be used for this purpose.

One good reason to stick with Django models, is the automatic creation of an editor
to change the content of thereof. By using the Django app
[django-entangled](https://pypi.org/project/django-entangled/), we can replace the
Django model by a Django form. This means that instead of declaring our data structure
inside a model, we now define it inside a form. This is a very similar approach
and gives the same result (in the sense of a dynamically created editor for our fields). 

## Proposal of change

So what I would like to propose is to add one JSONField to all models of the
django-CMS ecosystem which might be extended. This simplifies the creation of
third party apps, and reduces the number of database `JOIN`s and database migrations.

The app django-entangled can be used to create an editor for filling those
fields, but it shall remain an optional dependency.

In general I believe that by doing so, django-CMS will gain some benefits,
already offered by the Wagtail CMS (which relies much more on unstructured data).

### Clarification

I'm not opposing the classical approach for extending models provided by django-CMS
using a One-to-One relationship. In many situations this still might be the better
approach, but often it just is enough to keep that arbitrary data inside a JSONField
making the creation of third party extensions much simpler.

### Placeholder

One model which is not intended for extension is `cms.models.placeholdermodel.Placeholder`.
This model would also be a good candidate for adding such a JSONField. The model
`Placeholder` currently contains a field named `default_width`, which according to
my experience is never used. 

### Naming

Such an additional JSONField will hold arbitrary data with no relevance to django-CMS
itself. In the case of the `CMSPlugin` it just will act as additional context to the
rendering template. We therefore shall name it in a very generic way. Here are some
proposals:

* `context`: I'm a bit reluctant on this name, because it can cause confusion with other variables named like this.
* `extra_context`: Better, but too verbose IMO.
* `extra`, `data` or `extra_data`: Too generic, too boring.
* `scope`: This sounds good to me. Used btw. by the AngularJS framework to name "context".
* `glossary`: This currently is used in djangcms-cascade.  This sounds good to me.

Other proposals?

### Risk Assessment

Adding those fields doesn't change any functionality of the core. It only adds
a data bucket to third party apps. Therefore such a change in my opinion is free
of risks and could be introduced with a minor version number change.

**Comment by benzkji** [comment]
In general, I like the idea. Many times, I've done things with page templates, because I didnt wanted to mess with page extensions.

## Page
For me, the Page model would be the primary target of such a field. It would have almost no implications on existing projects (besides adding a new migration, under django-cms control). The new field would just be there, and be empty. Hopfully. As we all know, adding features always adds complexity ;-)

## Plugin
I can also see the point for the Plugin model. But, implications would be on a different level. Existing third party plugins (also all of the cms recommended plugins) would need to create a new migration, or developers would need to add them to MIGRATION_MODULES. It is well known that not all third party plugins are maintained very regularly. But I guess, with a clean and thoughtful strategy, it should be possible.

## Placeholder
I dont really see the point of a JSON field on the Placeholder, though? How would one fill this field? The `default_widht` is a very old legacy, as far as I know, and probably shoudl/could be deprecated/removed. Currently, one can add "extra_context" in `CMS_PLACEHOLDER_CONF` ( https://docs.django-cms.org/en/latest/reference/configuration.html?highlight=PLACEHOLDER_CONFIG#std-setting-CMS_PLACEHOLDER_CONF ), to have specific key/values for specific placeholders, this is already pretty ok, normally, for me. Open for arguments, though ;-)

## Naming
I would name it `extra_data`, as this is what it is? Scope is not well known in the Python world (at least not for context/data values ;_), a glossary is a glossary, context on it's own would confuse. So, maybe `extra_context`?

## Version
I agree with MINOR, as, technically, it is a backwards compatible change. Seeing the trouble/migrations/etc. it could cause, especially when adding the field to plugins, I would tend to make it a MAJOR (would be perfect for a 4.0).

**Reply by jrief** [reply]
Thanks Ben for taking the time to review this.

### Page & Title
You mentioned that a JSONField makes sense for model `Page` but the same applies to model `Title`. We have one `Title` object per page *and* language and some strings simply have to be kept translated. That's the reason the abstract model
`TitleExtension` exists.

### Placeholder
In my opinion model `Placeholder` makes a lot of sense. This is the starting point for every tree structure for plugins. I for instance would prefer to store the current column width for each Bootstrap's breakpoint in there. Then the root plugin can ask its placeholder, how wide that wrapping container/row/column/etc. is. This for djangocms-cascade would be very beneficial, since it keeps track of the widths of all containers, rows and columns it manages.

And that data could be perfectly filled using an additional attributes added to the templatetag `placeholder`. This keep things together, which belong together. Keeping that information inside the `settings.py`, makes the configuration very static. Imagine, if we start selling django-CMS themes together with their templates, then we would have to inform the customers to add special configuration dicts to their settings.py – not so customer friendly.

### Naming
Then I would prefer `extra_context` over `extra_data`, because, well – it's a context.

### Version
Are migrations such a trouble? Since the introduction of `south` and later Django migrations you normally don't even feel them. Did customers complain often about migrations?

**Reply by benzkji** [reply]
Yes, +1 for **Page & Title**!

**Placeholders**. For me there was never a need of this. If it's useful for certain cases, why not. Would those settings being filled with the "Placeholder Hamburger Menu", I guess?

**Migrations**, you are right, not one customer ever compained. But for developers, migrations can be painful. I consider myself experienced, but not only one time I have created and applied unwanted migrations in apps installed in site-packages. A thing to avoid (I can handle it, but less experienced peopl can go crazy!), and exactly what could happen if a third party plugin does not keep up with the cms, after the JSON release, no? For me, it happened most of the times because a package was already long time not maintained, and a change in Django caused it. Personally I would definitly put it in the 4.0 release.

**Reply by jrief** [reply]
Ok. If we can agree and there are no other objections, I will prepare a pull request for the version 4 branch.
Before that, I have to get familiar with the version 4 branch myself.
